_posts/2020-07-06-private-bastions.md

```diff
@@ -0,0 +1,179 @@
+---
+layout: post
+title: 'Private Bastion - I'll do what you want me to do'
+date: 2020-07-06
+tags: aws ssm vpc ssh ec2
+author: Paul Symons
+---
+
+This post and its title are inspired by Tina Turner and her 1984 album "Private Dancer". Themes of fortitude, defiance and resiliency echo throughout her life story and her works, yet it is her belief that we can find a better way, that resonates most strongly right now.
+
+> *May you always find a way to turn poison into medicine* <br/><br/>Tina Turner, 2019
+
+# Private Bastions - Using VPC Endpoints to teleport into your VPC
+
+[Bastion Hosts](https://en.wikipedia.org/wiki/Bastion_host) are a fairly well known and aged concept - that of a server that acts as a solitary, publicly accessible network access point, placing authorised users within the perimeter of an otherwise private network.
+
+The problem with any bastion host is the public attack surface that it leaves exposed. This can be mitigated to an extent by restricting IP ranges, disabling root logins, using private key passwords and the likes; yet the nature of the world dictates that a public bastion host can never be risk free.
+
+AWS Systems Manager released a feature in 2018 called *Session Manager*, allowing the ability to connect directly to EC2 hosts without the need to expose any ports on those hosts (for example, port 22 for SSH). There are two ways you can make this work:
+
+* Place your EC2 in a public subnet with a public IPV4 address
+* Place your EC2 in a private subnet and create VPC endpoints to SSM related services
+
+Today we'll focus on the second strategy.
+
+## Use Cases
+
+Here's a couple of scenarios to explain why you might want to use Session Manager to access resources in private subnets.
+
+The first scenario is that you deploy an application consisting of microservices and a database to private subnets, made public by a load balancer in a public subnet. If you want the ability to test these services within the network, a private bastion might be a good way to do that:
+
+![Scenario 1](/img/blog/ssm-private-terminal/scenario1.png)
+
+Another use case is that you have a hybrid network consisting of AWS resources and on-premise networks, perhaps by interconnected by a Transit Gateway attachment. In such scenarios, you may be forbidden from placing EC2 resources in public subnets or you may not even have an Internet Gateway at all:
+
+![Scenario 2](/img/blog/ssm-private-terminal/scenario2.png)
```

Comment:
Your architecture diagrams just use `private` networks.  This conflicts with your example below where you're using `isolated`.  Can you make them consistent?

Public - what it says
Private - internet egress via NAT Gateway
Isolated - private only.

Isolated is great for your example to demonstrate that it's truly internal only.

Reply:
Thanks @temyers very useful feedback. Have changed diagrams and some more text to add consistency